data/src/main/java/co/nimblehq/compose/crypto/data/ResponseMapping.kt

```diff
@@ -0,0 +1,39 @@
+package co.nimblehq.compose.crypto.data
+
+import co.nimblehq.compose.crypto.data.model.response.ErrorResponse
+import co.nimblehq.compose.crypto.data.service.JsonApiException
+import co.nimblehq.compose.crypto.data.service.UnknownException
+import co.nimblehq.compose.crypto.data.service.providers.MoshiBuilderProvider
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.flow
+import retrofit2.Response
+import kotlin.experimental.ExperimentalTypeInference
+
+@OptIn(ExperimentalTypeInference::class)
+fun <T> flowTransform(@BuilderInference block: suspend FlowCollector<T>.() -> Response<T>) = flow {
+    val response = block()
+    val body = response.body()
+    if (response.isSuccessful && body != null) {
+        emit(body)
+    } else {
+        error(mapError(response))
```

Comment:
@hoangnguyen92dn I just figured out, we should use `throw` instead of `error` here. `error` will wrap and throw the message as Any into a new `IllegalStateException` class, instead of the raw throwable.

<img width="760" alt="image" src="https://user-images.githubusercontent.com/16315358/187180867-06e6c9f9-b9ea-4f65-ac9e-13c0106815b7.png">


```suggestion
        throw mapError(response)
```